test/gtest/uct/ib/test_sockaddr.cc

```diff
@@ -855,6 +831,214 @@ UCS_TEST_P(test_uct_cm_sockaddr, connect_client_to_server_reject_with_delay)
 UCT_INSTANTIATE_SOCKADDR_TEST_CASE(test_uct_cm_sockaddr)
 
 
+class test_uct_cm_sockaddr_enhance_tests : public test_uct_cm_sockaddr {
+public:
+    test_uct_cm_sockaddr_enhance_tests() : m_enhance_clients_num(0),
+                                           m_ep_init_disconnect_cnt(0) {
+    }
+
+    void init() {
+        test_uct_cm_sockaddr::init();
+
+        m_enhance_clients_num = ucs_max(2, 100 / ucs::test_time_multiplier());
+    }
+
+    int get_ep_index (uct_ep_h ep) {
+        for (int i = 0; i < (2 * m_enhance_clients_num); i++) {
+            if (m_all_eps[i].ep == ep) {
+                return i;
+            }
+        }
+
+        return -1;
+    }
+
+    void common_enhance_test_disconnect(uct_ep_h ep) {
+        int index;
+
+        index = get_ep_index(ep);
+        EXPECT_GE(index, 0);
+        EXPECT_LT(index, (2 * m_enhance_clients_num));
+
+        pthread_mutex_lock(&m_lock);
+        m_all_eps[index].state |= TEST_CM_EP_FLAG_DISCONNECT_CB_INVOKED;
+        pthread_mutex_unlock(&m_lock);
+
+        if (!(m_all_eps[index].state & TEST_CM_EP_FLAG_DISCONNECT_INITIATOR)) {
+            ASSERT_UCS_OK(uct_ep_disconnect(ep, 0));
+        } else {
+            pthread_mutex_lock(&m_lock);
+            m_ep_init_disconnect_cnt--;
```

Comment:
is it safety to update state and counter under different lock-sections?

Reply:
i think that it is in this case. do you see a case where it's not?

Reply:
I don't but it means that we can't check the state and the counter in the same test. (looks true for now)